packages/tokens/src/rwa/compliance/mod.rs

```diff
@@ -0,0 +1,267 @@
+use soroban_sdk::{contracterror, contracttype, Address, Env, String, Symbol, Vec};
+
+pub mod storage;
+
+#[cfg(test)]
+mod test;
+
+/// Hook types for modular compliance system.
+///
+/// Each hook type represents a specific event or validation point
+/// where compliance modules can be executed.
+#[contracttype]
+#[derive(Clone, Debug, Eq, PartialEq, PartialOrd, Ord)]
+pub enum HookType {
+    /// Called after tokens are successfully transferred from one wallet to
+    /// another. Modules registered for this hook can update their state
+    /// based on transfer events.
+    Transfer,
+
+    /// Called after tokens are successfully created/minted to a wallet.
+    /// Modules registered for this hook can update their state based on minting
+    /// events.
+    Created,
+
+    /// Called after tokens are successfully destroyed/burned from a wallet.
+    /// Modules registered for this hook can update their state based on burning
+    /// events.
+    Destroyed,
+
+    /// Called during transfer validation to check if a transfer should be
+    /// allowed. Modules registered for this hook can implement transfer
+    /// restrictions. This is a READ-only operation and should not modify
+    /// state.
+    CanTransfer,
+}
+
+/// Trait for implementing custom compliance logic to RWA tokens.
+///
+/// [`Compliance`] trait is not expected to be an extension to a RWA smart
+/// contract, but it is a separate contract on its own. This design allows it to
+/// be shared across many RWA tokens. Note that, there is no `RWA` bound on the
+/// [`Compliance`] trait:
+///
+/// ```rust, ignore
+/// pub trait Compliance       // ✅
+/// pub trait Compliance: RWA  // ❌
+/// ```
+pub trait Compliance {
```

Comment:
Each of `transferred()`, `created()` and `destroyed()` can be state changing, so we need to add an additional parameter `token: Address` on which to call `require_auth()` and check whether it was bound to `Compliance`.

Reply:
let's continue this discussion from Slack as well